File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    var answer = 0;
    
    const start = ['[','(','{'];
    const end = [']',')','}'];
    
    const len = s.length;
    for(let i =0; i<len; i++) {
        
        const tmpStr = s.slice(i) + s.slice(0, i);
        
        let flag = true; 
        let arr = [];

        for(const item of tmpStr) {
            if(start.includes(item)) arr.push(item);
            else {
                const idx = end.indexOf(item);

                if(arr.pop() !== start[idx]) {
                    flag = false;
                    break;
                }
            }
        }
        
        if(flag && arr.length ===0) answer +=1;
    }
    
    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(string s) {
    int answer = 0;
    
    vector<char>start = {'[', '(', '{'};
    vector<char>last = {']',')','}'};
    
    int len = s.length();
    for(int i=0; i< len; i++) 
    {
        string tmpStr = s.substr(i,len- i) + s.substr(0, i);
        
        stack <char> pouch;
        bool flag = true;
        for(auto& item : tmpStr)
        {
            auto startIdx = find(start.begin(), start.end(), item);
            
            
            if(startIdx != start.end()) pouch.push(item);
            else 
            {
                if(pouch.empty())
                {
                    flag = false;
                    break;
                }
                
                auto endIdx = find(last.begin(), last.end(), item) - last.begin();
                
                if(pouch.top() != start[endIdx]) 
                {
                    flag = false;
                    break;
                } 
                else pouch.pop();
            }
        }
        
        if(flag && pouch.empty()) answer +=1;
    }
    
    return answer;
}